// talkback/features/error-rate.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var options_1 = __importDefault(require("../options"));
var logger_1 = require("../logger");
var ErrorRate = /** @class */ (function () {
    function ErrorRate(options) {
        this.options = options;
    }
    ErrorRate.prototype.shouldSimulate = function (req, tape) {
        var globalErrorRate = typeof (this.options.errorRate) === 'number' ? this.options.errorRate : this.options.errorRate(req);
        var errorRate = tape && tape.meta.errorRate !== undefined ? tape.meta.errorRate : globalErrorRate;
        options_1.default.validateErrorRate(errorRate);
        var random = Math.random() * 100;
        return random < errorRate;
    };
    ErrorRate.prototype.simulate = function (req) {
        logger_1.logger.log.info("Simulating error for ".concat(req.url));
        return {
            status: 503,
            headers: { 'content-type': ['text/plain'] },
            body: Buffer.from("talkback - failure injection")
        };
    };
    return ErrorRate;
}());
exports.default = ErrorRate;
