// talkback/logger.js
"use strict";
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.initializeLogger = exports.logger = exports.NoopLogger = exports.ConsoleLogger = void 0;
var ConsoleLogger = /** @class */ (function () {
    function ConsoleLogger(options) {
        this.options = options;
        if (this.options.debug) {
            this.debug("DEBUG mode active");
        }
    }
    ConsoleLogger.prototype.info = function (message) {
        if (!this.options.silent || this.options.debug) {
            console.log(this.formatMessage(message, "INFO"));
        }
    };
    ConsoleLogger.prototype.debug = function (message) {
        if (this.options.debug) {
            console.debug(this.formatMessage(message, "DEBUG"));
        }
    };
    ConsoleLogger.prototype.error = function (message) {
        var optionalParameters = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            optionalParameters[_i - 1] = arguments[_i];
        }
        console.error.apply(console, __spreadArray([this.formatMessage(message, "ERROR")], optionalParameters, false));
    };
    ConsoleLogger.prototype.formatMessage = function (message, level) {
        var now = new Date();
        var formattedNow = now.toISOString();
        var messageString;
        if (typeof message == "object") {
            messageString = JSON.stringify(message);
        }
        else {
            messageString = message;
        }
        return "".concat(formattedNow, " [").concat(this.options.name, "] [").concat(level, "] ").concat(messageString);
    };
    return ConsoleLogger;
}());
exports.ConsoleLogger = ConsoleLogger;
var NoopLogger = /** @class */ (function () {
    function NoopLogger() {
    }
    NoopLogger.prototype.debug = function (message) {
    };
    NoopLogger.prototype.error = function (message) {
        var optionalParameters = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            optionalParameters[_i - 1] = arguments[_i];
        }
    };
    NoopLogger.prototype.info = function (message) {
    };
    return NoopLogger;
}());
exports.NoopLogger = NoopLogger;
var logger = {
    log: new NoopLogger()
};
exports.logger = logger;
function initializeLogger(options) {
    logger.log = new ConsoleLogger(options);
}
exports.initializeLogger = initializeLogger;
