// talkback/tape-matcher.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var content_encoding_1 = __importDefault(require("./utils/content-encoding"));
var media_type_1 = __importDefault(require("./utils/media-type"));
var logger_1 = require("./logger");
var isEqual = require("lodash/isEqual");
var TapeMatcher = /** @class */ (function () {
    function TapeMatcher(tape, options) {
        this.tape = tape;
        this.options = options;
    }
    TapeMatcher.prototype.sameAs = function (otherTape) {
        var otherReq = otherTape.req;
        var req = this.tape.req;
        if (!this.isSameUrl(req, otherReq)) {
            return false;
        }
        if (!this.isSameMethod(req, otherReq)) {
            return false;
        }
        if (!this.isSameHeaders(req, otherReq)) {
            return false;
        }
        return this.options.ignoreBody || this.isSameBody(req, otherReq);
    };
    TapeMatcher.prototype.isSameBody = function (req, otherReq) {
        var mediaType = new media_type_1.default(req);
        var contentEncoding = new content_encoding_1.default(req);
        var sameBody;
        if (contentEncoding.isUncompressed() && mediaType.isJSON() && req.body.length > 0 && otherReq.body.length > 0) {
            var parsedReqBody = JSON.parse(req.body.toString());
            var parsedOtherReqBody = JSON.parse(otherReq.body.toString());
            sameBody = isEqual(parsedReqBody, parsedOtherReqBody);
        }
        else {
            sameBody = req.body.equals(otherReq.body);
        }
        if (!sameBody) {
            if (!this.options.bodyMatcher) {
                logger_1.logger.log.debug("Not same BODY ".concat(req.body, " vs ").concat(otherReq.body));
                return false;
            }
            var bodyMatches = this.options.bodyMatcher(this.tape, otherReq);
            if (!bodyMatches) {
                logger_1.logger.log.debug("Not same bodyMatcher ".concat(req.body, " vs ").concat(otherReq.body));
                return false;
            }
        }
        return true;
    };
    TapeMatcher.prototype.isSameHeaders = function (req, otherReq) {
        var currentHeadersLength = Object.keys(req.headers).length;
        var otherHeadersLength = Object.keys(otherReq.headers).length;
        var sameNumberOfHeaders = currentHeadersLength === otherHeadersLength;
        if (!sameNumberOfHeaders) {
            logger_1.logger.log.debug("Not same #HEADERS ".concat(JSON.stringify(req.headers), " vs ").concat(JSON.stringify(otherReq.headers)));
            return false;
        }
        var headersSame = true;
        Object.keys(req.headers).forEach(function (k) {
            var entryHeader = req.headers[k];
            var header = otherReq.headers[k];
            headersSame = headersSame && entryHeader === header;
        });
        if (!headersSame) {
            logger_1.logger.log.debug("Not same HEADERS values ".concat(JSON.stringify(req.headers), " vs ").concat(JSON.stringify(otherReq.headers)));
            return false;
        }
        return true;
    };
    TapeMatcher.prototype.isSameMethod = function (req, otherReq) {
        var sameMethod = req.method === otherReq.method;
        if (!sameMethod) {
            logger_1.logger.log.debug("Not same METHOD ".concat(req.method, " vs ").concat(otherReq.method));
            return false;
        }
        return true;
    };
    TapeMatcher.prototype.isSameUrl = function (req, otherReq) {
        var sameURL = req.url === otherReq.url;
        if (!sameURL) {
            if (!this.options.urlMatcher) {
                logger_1.logger.log.debug("Not same URL ".concat(req.url, " vs ").concat(otherReq.url));
                return false;
            }
            var urlMatches = this.options.urlMatcher(this.tape, otherReq);
            if (!urlMatches) {
                logger_1.logger.log.debug("Not same urlMatcher ".concat(req.url, " vs ").concat(otherReq.url));
                return false;
            }
        }
        return true;
    };
    return TapeMatcher;
}());
exports.default = TapeMatcher;
