// talkback/tape-store.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var fs = require("fs");
var path = require("path");
var JSON5 = require("json5");
var mkdirp = require("mkdirp");
var tape_1 = __importDefault(require("./tape"));
var tape_matcher_1 = __importDefault(require("./tape-matcher"));
var tape_renderer_1 = __importDefault(require("./tape-renderer"));
var logger_1 = require("./logger");
var TapeStore = /** @class */ (function () {
    function TapeStore(options) {
        this.path = path.normalize(options.path + "/");
        this.options = options;
        this.tapes = [];
    }
    TapeStore.prototype.load = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        mkdirp.sync(this.path);
                        return [4 /*yield*/, this.loadTapesAtDir(this.path)];
                    case 1:
                        _a.sent();
                        logger_1.logger.log.info("Loaded ".concat(this.tapes.length, " tapes from ").concat(this.path));
                        return [2 /*return*/];
                }
            });
        });
    };
    TapeStore.prototype.loadTapesAtDir = function (directory) {
        return __awaiter(this, void 0, void 0, function () {
            var items, i, filename, fullPath, stat, data, raw, tape, e_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        items = fs.readdirSync(directory);
                        i = 0;
                        _a.label = 1;
                    case 1:
                        if (!(i < items.length)) return [3 /*break*/, 8];
                        filename = items[i];
                        fullPath = "".concat(directory).concat(filename);
                        stat = fs.statSync(fullPath);
                        if (!!stat.isDirectory()) return [3 /*break*/, 6];
                        _a.label = 2;
                    case 2:
                        _a.trys.push([2, 4, , 5]);
                        data = fs.readFileSync(fullPath, "utf8");
                        raw = JSON5.parse(data);
                        return [4 /*yield*/, tape_1.default.fromStore(raw, this.options)];
                    case 3:
                        tape = _a.sent();
                        tape.path = filename;
                        this.tapes.push(tape);
                        return [3 /*break*/, 5];
                    case 4:
                        e_1 = _a.sent();
                        logger_1.logger.log.error("Error reading tape ".concat(fullPath), e_1.message);
                        return [3 /*break*/, 5];
                    case 5: return [3 /*break*/, 7];
                    case 6:
                        this.loadTapesAtDir(fullPath + "/");
                        _a.label = 7;
                    case 7:
                        i++;
                        return [3 /*break*/, 1];
                    case 8: return [2 /*return*/];
                }
            });
        });
    };
    TapeStore.prototype.find = function (newTape) {
        var _this = this;
        var foundTape = this.tapes.find(function (t) {
            logger_1.logger.log.debug("Comparing against tape ".concat(t.path));
            return new tape_matcher_1.default(t, _this.options).sameAs(newTape);
        });
        if (foundTape) {
            foundTape.used = true;
            logger_1.logger.log.info("Found matching tape for ".concat(newTape.req.url, " at ").concat(foundTape.path));
            return foundTape;
        }
    };
    TapeStore.prototype.save = function (tape) {
        return __awaiter(this, void 0, void 0, function () {
            var tapePath, fullFilename, tapeRenderer, toSave;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        tape.new = true;
                        tape.used = true;
                        tapePath = tape.path;
                        if (tapePath) {
                            fullFilename = path.join(this.path, tapePath);
                        }
                        else {
                            // If the tape doesn't have a path then it's new
                            this.tapes.push(tape);
                            fullFilename = this.createTapePath(tape);
                            tape.path = path.relative(this.path, fullFilename);
                        }
                        logger_1.logger.log.info("Saving request ".concat(tape.req.url, " at ").concat(tape.path));
                        tapeRenderer = new tape_renderer_1.default(tape);
                        return [4 /*yield*/, tapeRenderer.render()];
                    case 1:
                        toSave = _a.sent();
                        fs.writeFileSync(fullFilename, JSON5.stringify(toSave, null, 4));
                        return [2 /*return*/];
                }
            });
        });
    };
    TapeStore.prototype.currentTapeId = function () {
        return this.tapes.length;
    };
    TapeStore.prototype.hasTapeBeenUsed = function (tapeName) {
        return this.tapes.some(function (t) { return t.used && t.path === tapeName; });
    };
    TapeStore.prototype.resetTapeUsage = function () {
        return this.tapes.forEach(function (t) { return t.used = false; });
    };
    TapeStore.prototype.createTapePath = function (tape) {
        var currentTapeId = this.currentTapeId();
        var tapePath = "unnamed-".concat(currentTapeId, ".json5");
        if (this.options.tapeNameGenerator) {
            tapePath = this.options.tapeNameGenerator(currentTapeId, tape);
        }
        var result = path.normalize(path.join(this.options.path, tapePath));
        if (!result.endsWith(".json5")) {
            result = "".concat(result, ".json5");
        }
        var dir = path.dirname(result);
        mkdirp.sync(dir);
        return result;
    };
    return TapeStore;
}());
exports.default = TapeStore;
