// talkback/utils/media-type.js
"use strict";
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.jsonTypes = void 0;
var headers_1 = __importDefault(require("./headers"));
var contentTypeParser = require("content-type");
var equals = function (to) { return function (contentType) { return to == contentType; }; };
exports.jsonTypes = [
    equals("application/json"),
    equals("application/x-amz-json-1.0"),
    equals("application/x-amz-json-1.1"),
    function (contentType) { return contentType.startsWith("application/") && contentType.endsWith("+json"); }
];
var humanReadableContentTypes = __spreadArray([
    equals("application/javascript"),
    equals("text/css"),
    equals("text/html"),
    equals("text/javascript"),
    equals("text/plain")
], exports.jsonTypes, true);
var MediaType = /** @class */ (function () {
    function MediaType(htmlReqRes) {
        this.htmlReqRes = htmlReqRes;
    }
    MediaType.prototype.isHumanReadable = function () {
        var contentType = this.contentType();
        if (!contentType) {
            return false;
        }
        return humanReadableContentTypes.some(function (comparator) { return comparator(contentType); });
    };
    MediaType.prototype.isJSON = function () {
        var contentType = this.contentType();
        if (!contentType) {
            return false;
        }
        return exports.jsonTypes.some(function (comparator) { return comparator(contentType); });
    };
    MediaType.prototype.contentType = function () {
        var contentTypeHeader = headers_1.default.read(this.headers(), "content-type");
        if (!contentTypeHeader) {
            return null;
        }
        var parsedContentType = contentTypeParser.parse(contentTypeHeader);
        return parsedContentType.type;
    };
    MediaType.prototype.headers = function () {
        return this.htmlReqRes.headers;
    };
    return MediaType;
}());
exports.default = MediaType;
